### DISCUSSION TITLE
Expectation value of site occupation is not translationally invariant

### DISCUSSION THREAD

**Discussion by KoutaDagnino** [QuSpin/QuSpin#442] [Q&A] [answered]
Dear QuSpin team,
I was wondering if the following method to evaluate expectation values for site occupation operators is correct.

Take for example this code slightly modified from https://github.com/weinbe58/QuSpin/blob/master/examples/notebooks/BHM.ipynb.

`from quspin.operators import hamiltonian # Hamiltonians and operators
from quspin.basis import boson_basis_1d # Hilbert space boson basis
import numpy as np # generic math functions
import matplotlib.pyplot as plt # to plot
#
##### define model parameters #####
L=10 # system size
N=5
J=1.0 # hopping
U=np.sqrt(2.0) # interaction
mu=0.0 # chemical potential
#
##### construct Bose-Hubbard Hamiltonian #####

basis = boson_basis_1d(L,Nb=N)
print(basis.Ns)

hop=[[-J,i,(i+1)%L] for i in range(L)] #PBC
interact=[[0.5*U,i,i] for i in range(L)] # U/2 \sum_j n_j n_j
pot=[[-mu-0.5*U,i] for i in range(L)] # -(\mu + U/2) \sum_j j_n

static=[['+-',hop],['-+',hop],['n',pot],['nn',interact]]
dynamic=[]

H=hamiltonian(static,dynamic,basis=basis,dtype=np.float64,check_symm=False)

E,states=H.eigh()
##### construct site occupation operators #####
op2_static=[["n", [[1,3]]]]
op3_static=[["n", [[1,5]]]]

op2 = hamiltonian(op2_static,dynamic,basis=basis, check_symm=False)
op3 = hamiltonian(op3_static,dynamic,basis=basis, check_symm=False)

##### compute expectation values #####
Energy1=[]
Expectation1=[]
Expectation2=[]
Energy1.extend(E)

deg=states.shape[1]

for i in range(0,deg):
    expt2 = op2.expt_value(states[:,i])
    Expectation1.extend([expt2])
    expt3 = op3.expt_value(states[:,i])
    Expectation2.extend([expt3])

Expectation1=np.array(Expectation1) 
Expectation2=np.array(Expectation2) 

Energy1=np.array(Energy1) 

##### Plotting #####
plt.scatter(Energy1,Expectation1,marker="o",s=2,color="r",alpha=0.3,label="L=12")
plt.scatter(Energy1,Expectation2,marker="o",s=2,color="k",alpha=0.3,label="L=12")
plt.show()`

When I plot the expectation value for the occupation of sites 3 and 5, they don't seem to coincide. However shouldn't they be the same due to translational invariance of the system?

**Comment by mgbukov** [comment, marked as answer]
Hi! 

to answer your question directly: yes, expectation values in a transl. invariant system should be the same. This is because the eigenstates of `H` _*can*_ be chosen to be simultaneous e'states of the translation operator. However, in practice if you have degeneracies, any linear combination in the degenerate manifold will be a valid e'vector; now, if the degenerate manifold is spanned by vectors of two different quantum numbers of the transl. op., their superposition will not be an e'state of the transl. op and hence this e'state will not be translation invariant. This is because when diagonalizing the matrix `H` we don't pass any information about the translation op to _simultaneously_ diagonalize both so the algorithm picks any e'states that do the job. 

A quick check for you on the first three e'states (for L=6) gives:
```
print(Energy1[:3]) # [-8.13863961 -6.47006376 -6.47006376]
print(Expectation1[:3]) # [0.83333333+0.j 0.76899255+0.j 0.89767411+0.j]
print(Expectation2[:3]) # [0.83333333+0.j 0.86579226+0.j 0.80087441+0.j]
```
As you see, if the e'values are unique, the expectation values match; when there are degeneracies, the e'states appear as linear combinations which break transl symmetry (as discussed above). 


A few general remarks about your quspin code:
* do not turn off the symmetry check unless you are 100% you know what you're doing. E.g., had you specified `kblock=0` in the basis, the code would've gone thru but the operator $n_j$ breaks transl. invariance. If you want to use transl invariance in `basis` and still measure local operators which break trasl. inv, check out [this example](https://weinbe58.github.io/QuSpin/examples/example7.html#example7-label).
* you do not need to loop over states when using `expt_value`: check out the doc, there is a way to apply this function on a bunch of states simultaneously
* you may want to explicitly use `expt_value().real` to remove the imaginary part.

**Reply by KoutaDagnino** [reply]
Thank you so much! So diagonalizing the matrix in a symmetry block should solve the issue right?

**Reply by mgbukov** [reply]
precisely! But your local operators $n_j$ are not well-defined within these symmetry sectors (they break the transl symmetry); so check out the example in the first bullet point above.

**Reply by KoutaDagnino** [reply]
Thank you for your help!
